sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -675,6 +676,29 @@ object PushPredicateThroughGenerate extends Rule[LogicalPlan] with PredicateHelp
 }
 
 /**
+ * Push [[Filter]] operators through [[Aggregate]] operators. Parts of the predicate that reference
+ * attributes which are subset of group by attribute set of [[Aggregate]] will be pushed beneath, 
+ * and the rest should remain above.
+ */
+object PushPredicateThroughAggregate extends Rule[LogicalPlan] with PredicateHelper {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case filter @ Filter(condition,
+    aggregate @ Aggregate(groupingExpressions, aggregateExpressions, grandChild)) =>
```

Comment:
nit: indentation.  I'm not sure we have a strict rule, but no indent is kinda hard to follow.  I'd probably try to make it a tree if it fits?

```scala
case filter @ Filter(condition,
     aggregate @ Aggregate(groupingExpressions, aggregateExpressions, grandChild)) =>
```

or just 4 space indent?